Checkpoint3 Submission

lewisemm opened the pull request.

lewisemm committed:
Initial Commit

lewisemm committed:
(feature) Create initial tests for TDD

lewisemm committed:
(chore) Create ".gitignore" file

lewisemm committed:
(feature) Create models for BucketList and Item objects

lewisemm committed:
(chore) Update "requirements.txt" file

lewisemm committed:
(feature) Add more tests for TDD

lewisemm committed:
(feature) Create django project "checkpoint3" and app "api"

lewisemm committed:
(chore) Update requirements file with "dj_database_url" package

lewisemm committed:
(feature) Configure project's settings.py file

- Include api app in INSTALLED_APPS
- Configure appropriate time one
- Use package dj_database_url to retrieve database configurations from DATABASE_URL environment variable

lewisemm committed:
(feature) Create app's models

lewisemm committed:
(chore) Update .gitignore file.

- Ignores migrations

lewisemm committed:
(feature) Add BucketList and Item to the admin interface

lewisemm committed:
(chore) Add migrations and update ".gitignore" file to ignore autoenv's ".env" file

lewisemm committed:
(feature) Add foreign key to items model

lewisemm committed:
(feature) Wire up djangorestframework

- Added configurations in settings.py file
- Added routing to api app
- Created viewsets and serializers
- Configured viewset urls to api.urls file

lewisemm committed:
(chore) Add psycopg2 package to file

lewisemm committed:
(feature) Modify models

- Create primary field using models.AutoField for auto-increment capability
- Create constraints for model fields (blank=True)
- Create migrations

lewisemm committed:
(feature) Customize bucketlist creation process

- Defined create method in BucketListViewSet to call "create_bucketlist" method in models.BucketList
- This method creates a bucketlist and inserts the currently logged in user as the creator of the BucketList

lewisemm committed:
(feature) Refactor Serializers

- Remove 'url' from fields tuple to eliminate error
- Add 'bucketlist' field to fields tuple in serializer to provide model Item's foreign key on viewset.

lewisemm committed:
(chore) Add ipdb package to requirements.txt file

lewisemm committed:
(feature) Create logic to display nested BucketList

lewisemm committed:
(feature) Use router from drf-nested-routers to create nested REST API routes

lewisemm committed:
(chore) Remove django admin route

lewisemm committed:
(feature) Create "list" and "retrieve" methods in ViewSets

- Created retrieve and list methods in ItemViewSet to filter out items that aren't
  posted under the bucketlist whose id is specified in the URL

lewisemm committed:
(chore) Apply PEP8 and PEP257 conventions

lewisemm committed:
(feature) Create User serializer, viewset and access url

lewisemm committed:
(feature) Create custom permission to grant write access only to owners

lewisemm committed:
(chore) Remove global authentication permissions

lewisemm committed:
(feature) Tweak permissions.py to deny Item write access

- Permissions refactored to deny write access to currently logged in user if he/she doesn't own parent BucketList

lewisemm committed:
(feature) Create and configure django templates and Materialize CSS for landing page

lewisemm committed:
(feature) Create login page  and include its url

lewisemm committed:
(feature) Use Angular JS to retrieve BucketList data

lewisemm committed:
(fix) Configure Angular to use "||" instead of "{{}}" to avoid conflicts with Django templates

lewisemm committed:
(feature) Add fixed header navbar and fixed footer

lewisemm committed:
(feature) Use materialize sliders on index page

lewisemm committed:
(feature) Use Angular JS to display materialize cards containing Bucketlist items

lewisemm committed:
(feature) Refactor login page to allow both sign up and sign in (through tabs)

lewisemm committed:
(feature) Create login controller to handle user login

- Move ng-app declaration from "bucketlist.html" to "base.html" to allow universal app coverage
- Add ng-click to Sign In button at "login.html"
- Declare "LoginController.js" as a service on "login.html"

lewisemm committed:
(chore) Install djangorestframework_jwt and configure settings and url

- Also updated requirements.txt with newly installed packages

lewisemm committed:
(feature) Create Angular app, services and controllers

lewisemm committed:
(fix) Move templates from templates folder to static/views.

- Create css to establish a  sticky footer at base.html

lewisemm committed:
(fix) Include more fields in the field tuple of the serializer to display responses as specified

lewisemm committed:
(feature) Add route to index page and create view to render html file of index page

lewisemm committed:
(fix) Include api/static/views as a template directory

lewisemm committed:
(chore) Ignore .DS_Store file

lewisemm committed:
(feature) Create permissions to deny unsafe methods to non-owners

- Apply custom permissions to Bucketlist and Item viewsets

lewisemm committed:
(chore) Fix indentation

lewisemm committed:
(feature) Add read-only fields that are required for get requests but not required for post/put requests

lewisemm committed:
(feature) Add ngCookies dependency to enable storage of authenticatio token

- Create routes to view bucketlist items and edit them as well

lewisemm committed:
(feature) Restructure BucketlistFactory service.

- Restructure logic to handle GET, PUT, POST and DELETE methods
- Create Item resource to handle item addition to bucketlist
- Create ItemDetails resource with edit and delete options

lewisemm committed:
(feature) Restructure scope bucketlists initialization to apply new BucketlistFactory structure

- Create scoped methods addItem, editeItem and deleteitem to perform POST, PUT and DELETE operations respectively

lewisemm committed:
(feature) Create logic to attach authentication token to a cookie under the key name 'Authorization'

lewisemm committed:
(feature) Add materialize modal dialog to aid in creattion of a new bucketlist

- Bind 'View bucketlist items', 'edit' and 'delete' buttons to BucketlistController's 'viewItems', 'edit' and 'del' methods respectively

lewisemm committed:
(feature) Create logic to display all items in selected bucketlist.

- Include materialize modal dialog to aid in addition of items to bucketlist
- Bind 'Create Item' button in materialize modal to addItem in BucketlitDetailController

lewisemm committed:
(chore) Alter index page to display images in static/images folder

lewisemm committed:
Create base.html to be extended by the rest of the app's pages. 'base.html' also contains Angular's ng-view directive

lewisemm committed:
(feature) Create ItemController that initializes scope object item with data from the database on the view

- Create view 'item_edit' to facillitate updating of bucketlist item details

lewisemm committed:
(feature) Create service to attach an authentication token from the cookie to every request's header for authentication purposes when accessing protected routes

lewisemm committed:
(fix) Temporarily remove faulty permissions from viewsets to allow development on front end

- Add create method to ItemViewSet to guarantee the creation of items with an associated bucketlist

lewisemm committed:
(feature) Create a paginator class to control how many results are returned to the client

- Declared the paginator class in the BucketlistViewSet and ItemViewSet

lewisemm committed:
(feature) Install and configure django-rest-swagger

lewisemm committed:
(fix) Refactor service code to expect an object instead of an array. Also refactor view at ng-repeat to iterate through the result object

lewisemm committed:
(fix) Refactor BucketlistViewSet get_queryset to allow searching of bucketlist by name if query parameter is provided

- Also fix viewset to return appropriate results when a bucketlist id is specified in the URL

lewisemm committed:
(feature) Refine View. Create validation logic and display error  messages

lewisemm committed:
(chore) Update header, footer and app theme colours

lewisemm committed:
(feature) Create sign up functionality and bind this to sign up view

lewisemm committed:
(feature) Refine bucketlist page. Implement bucketlist search (by name) functionality.

lewisemm committed:
(feature) Refine item edit UI. Implement promises to send appropriate messages when operation succeeds or fails

lewisemm committed:
Refine bucketlist details UI. Implement promises at delete items and create functions in BucketlistDetailController.

lewisemm committed:
(fix) Correct response messages

lewisemm committed:
(chore) Refine UI. Abstract styling to independent css files

lewisemm committed:
(feature) Use promises when fetching bucketlists. Display default image when there's no item to display.

lewisemm committed:
(feature) Implement working concept of pagination. Also include default view when bucketlist has no items.

lewisemm committed:
(feature) Implement logic to navigate to next page through the right chevron

lewisemm committed:
(feature) Implement logic to allow navigation to previous page using left chevron

lewisemm committed:
(feature) Create option for user to specify number of bucketlist items per page

lewisemm committed:
(chore) Apply PEP257 styling to paginator class. Remove console logs

lewisemm committed:
(chore) Remove CDNs. Install Materialize and Angular dependencies on host through bower.

lewisemm committed:
(feature) Replace JS confirm dialog before delete with more visually appealing SweetAlert dialog

lewisemm committed:
(chore) Change permission classes in Bucketlist viewset. Add status codes to responses and return serializer errors when data is not valid

lewisemm committed:
(feature) Create readme file containing project description, installation/run instructions.

lewisemm committed:
(fix) Fix malfunctioning permissions

lewisemm committed:
(feature) Create tests for login and bucketlist creation

lewisemm committed:
(feature) Test unauthenticated requests on '/bucketlist/' url

lewisemm committed:
(feature) Test read only permissions (if not owner) and write permissions (if owner) permissions on bucketlists

lewisemm committed:
(feature) Create tests for bucketlist items and their permissions

lewisemm committed:
(feature) Create tests to check for responses when invalid data is submitted

lewisemm committed:
(feature) Create update and destroy methods for the ItemViewSet.

- Also add permissions to ItemViewSet and confirm Bucketlist ownership before creating an item within one

lewisemm committed:
(feature) Add image resources for UI background displays

lewisemm committed:
(chore) Ignore .coverage file

lewisemm committed:
(feature) Create alternative database settings for TravisCI when TRAVIS_BUILD environment variable is available

lewisemm committed:
(feature) Create Travis CI configuration file

lewisemm committed:
(chore) Update requirements.txt file

lewisemm committed:
(feature) Add Travis CI, Coveralls and Quantified Code badges to readme

lewisemm committed:
(fix) Remove Travis CI's virtual env from test coverage. Fix coveralls badge that's not displaying properly

lewisemm committed:
(chore) Update requirements file

lewisemm committed:
(chore) Create coveralls configuration file. Update Travis CI config file

lewisemm committed:
(feature) Replace ordinary JS confirm dialog with more visually appealing SweetAlert dialog confirmation before delete operation

lewisemm committed:
(feature) Replace $http calls with $resource for SignUp and Login

- Create access_service.js containing User and NewUser resources to facilitate user login and signup respectively

lewisemm committed:
(fix) Refactor authorization_service to allow posts for endpoints that don't require authentication e.g. user sign up

lewisemm committed:
(chore) Add more details to readme

lewisemm committed:
(chore) Create bower.json and .bowerrc files

- bower.json contains bower components to be installed when "bower install" is run
- .bowerrc contains the path in which installed components will be stored on disk

lewisemm committed:
(chore) Ignore coveralls report file

lewisemm committed:
(chore) Create routes for Django REST framework login and Django REST Swagger documentation package

lewisemm committed:
(chore) Replace ordinary JavaScript prompt when edit is clicked with SweetAlert dialog.

- Add code to reload bucketlist items when the first item is added to a bucketlist
- Add code to reload bucketlists when one is deleted

lewisemm committed:
(feature) Add logic that appends user's username to the "Sign Out" button when logged in and defaults to "Sign Out" when logged out.

lewisemm committed:
(feature) Use Raleway font

lewisemm committed:
(fix) Comment django template statement to retain functionality and eliminate its display on pages

lewisemm committed:
(fix) Handle informative response errors at AuthorizationService. Also Redirect user to login when token expires

lewisemm committed:
(chore) Retrieve SECRET_KEY setting from environment variable

lewisemm committed:
(feature) Update item status to done when item checkbox is clicked

lewisemm committed:
(fix) Restrict Bucketlist and Item resources to authenticated users

lewisemm committed:
(fix) Add check for maximum page number when browsing paginated results

- Add logic to change text in header when user's token signature expires
- Increase div height to cater for additional field (created_by)

lewisemm committed:
(chore) Add Heroku dependency and config file

lewisemm committed:
(fix) Display username alongside SignOut button when user is logged in

- Check for presence of authorization token before showing SignIn button
- Remove the username cookie when the user signs out

lewisemm committed:
(fix) Fix and streamline various functionalities

- Remove angular filter and instead filter from backend through query parameter 'q'. Search returns bucketlists whose name matches/contains search string
- Remove authorization token and username information at LoginController to logout any user that may be logged in. Moved this logic from authorization_service
- Apply Angular's ng-class directive to highlight the current page number when scrolling through paginated results

lewisemm committed:
Merge pull request #1 from andela-lkabui/feature/name-search-pagination

(fix) Fix and streamline various functionalities

lewisemm committed:
Merge pull request #1 from andela-lkabui/feature/name-search-pagination

(fix) Fix and streamline various functionalities

lewisemm committed:
(fix) Deny read/write access permissions to bucketlists that do not belong to current user

- Restrict BucketListView queryset to bucketlists belonging to currently logged in user

lewisemm committed:
(fix) Reflect update on the view after the user clicks an item as done

lewisemm committed:
(fix) Invoke respective 'loadBucketlist(s)' to reload template data after delete and edit operations.

- Redirect to bucketlist details page when item is edited at the 'item_edit.html' view

lewisemm committed:
(fix) Maintain coveralls percentage coverage. Fine tune permissions. Refine docstrings.

lewisemm committed:
(fix) Fix Quantified Code Issues

lewisemm committed:
Merge pull request #2 from andela-lkabui/fix/permissions

Fix/permissions

lewisemm committed:
Merge pull request #3 from andela-lkabui/hotfix/refresh-after-edit

Hotfix/refresh after edit

lewisemm committed:
(chore) Update readme

lewisemm committed:
(chore) Update readme

lewisemm committed:
(fix) Fix 'no property "data" of undefined' error by returning an error response at the service.responseError function

- Remove console.log statements in controllers
- Remove 'created_by' field from bucketlist view because it's now irrelevant since one can only view the bucketlists they created

lewisemm committed:
Merge pull request #4 from andela-lkabui/hotfix/fix-interceptor-errors

Hotfix/fix interceptor errors

lewisemm committed:
(chore) Update readme

lewisemm committed:
Merge branch 'develop' into review

andela-anandaa commented on checkpoint3/urls.py:
```diff
@@ -0,0 +1,23 @@
+from django.conf.urls import url, include
+from django.contrib import admin
+
+from api import views
```
**Code Quality -- Nitpick**
This line is redundant.

andela-anandaa commented:
#### README / Comments
- The README is detailed and okay.
- You could have a proper name for your app, other than Checkpoint3

#### Directory Structure
- Separate the API (DRF/Django app) and the Django app (front-end).

#### Git / Commit Messages
- Some of your git commit messages was not well wrapped, see how to fix this [here](http://stackoverflow.com/questions/11023194/git-vim-automatically-wrap-long-commit-messages)

#### Code Quality
- Generally okay except for a few _severe_ issues raised by [Quantified Code](https://www.quantifiedcode.com/app/project/gh:andela-lkabui:checkpoint3/diff/1591b1f8cf96be0c70c7624801f9234d23d53294/03af4f0b8f512e29de1d91bd47129909129fe550/issues?groups=code_patterns%3A4Phzg6k7%3Af1,code_patterns%3A5ZdRTEsh%3Af6,code_patterns%3Akv4o9Hx2,code_patterns%3A7I7lGdCx%3Af3&severity=4)

#### Functionality
- Works as expected
- Fix the loading of the first bucket-list
- Fix token expiry

#### Testing
- Good: 100% test coverage.

#### UI/UX
- Need a favicon
- Fix session expiry to be more usable
- Call-to-action button
- "Item Status" => "Done"
- The rest of the nitpicks [here](https://drive.google.com/a/andela.co/folderview?id=0B8IqKJACNaFOUzk3Y2hRNFlEbDA&usp=sharing)

lewisemm committed:
(chore) Change app name

lewisemm committed:
Merge pull request #6 from andela-lkabui/chore/change-app-name

(chore) Change app name

lewisemm committed:
(fix) Resolve QuantifiedCode code quality issues

lewisemm committed:
(fix) Update Bucketlist model's created_by field

- created_by now refers to User class
- Adjust serializers, viewsets and tests to adapt to this change

lewisemm committed:
Merge pull request #7 from andela-lkabui/fix/code-quality

Address QuantifiedCode issues

andela-lkabui commented:
Good Morning Nandaa,

I trust your morning is great.

Having had the checkpoint3 review last week, I have addressed the nitpicks we identified during our session and I am ready to present the improvements.

I am hoping we can have some time before the end of this week (today if possible) to review the changes.

Please tell me if this works for you.
> On 18 Feb 2016, at 11:34, Anthony Nandaa <notifications@github.com> wrote:
> 
> README / Comments
> 
> The README is detailed and okay.
> You could have a proper name for your app, other than Checkpoint3
> Directory Structure
> 
> Separate the API (DRF/Django app) and the Django app (front-end).
> Git / Commit Messages
> 
> Some of your git commit messages was not well wrapped, see how to fix this here <http://stackoverflow.com/questions/11023194/git-vim-automatically-wrap-long-commit-messages>
> Code Quality
> 
> Generally okay except for a few severe issues raised by Quantified Code <https://www.quantifiedcode.com/app/project/gh:andela-lkabui:checkpoint3/diff/1591b1f8cf96be0c70c7624801f9234d23d53294/03af4f0b8f512e29de1d91bd47129909129fe550/issues?groups=code_patterns%3A4Phzg6k7%3Af1,code_patterns%3A5ZdRTEsh%3Af6,code_patterns%3Akv4o9Hx2,code_patterns%3A7I7lGdCx%3Af3&severity=4>
> Functionality
> 
> Works as expected
> Fix the loading of the first bucket-list
> Fix token expiry
> Testing
> 
> Good: 100% test coverage.
> UI/UX
> 
> Need a favicon
> Fix session expiry to be more usable
> Call-to-action button
> "Item Status" => "Done"
> The rest of the nitpicks here <https://drive.google.com/a/andela.co/folderview?id=0B8IqKJACNaFOUzk3Y2hRNFlEbDA&usp=sharing>
> —
> Reply to this email directly or view it on GitHub <https://github.com/andela-lkabui/checkpoint3/pull/5#issuecomment-185597691>.
> 

Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>

"The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
See what The Wall Street Journal <http://snip.ly/u9AI> has to say.

andela-lkabui commented:
I’ve just realised I did not send the link. Please find my repo here <https://github.com/andela-lkabui/checkpoint3/pull/12>.
> On 25 Feb 2016, at 11:34, Lewis Maina <lewis.maina@andela.com> wrote:
> 
> Good Morning Nandaa,
> 
> I trust your morning is great.
> 
> Having had the checkpoint3 review last week, I have addressed the nitpicks we identified during our session and I am ready to present the improvements.
> 
> I am hoping we can have some time before the end of this week (today if possible) to review the changes.
> 
> Please tell me if this works for you.
>> On 18 Feb 2016, at 11:34, Anthony Nandaa <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> README / Comments
>> 
>> The README is detailed and okay.
>> You could have a proper name for your app, other than Checkpoint3
>> Directory Structure
>> 
>> Separate the API (DRF/Django app) and the Django app (front-end).
>> Git / Commit Messages
>> 
>> Some of your git commit messages was not well wrapped, see how to fix this here <http://stackoverflow.com/questions/11023194/git-vim-automatically-wrap-long-commit-messages>
>> Code Quality
>> 
>> Generally okay except for a few severe issues raised by Quantified Code <https://www.quantifiedcode.com/app/project/gh:andela-lkabui:checkpoint3/diff/1591b1f8cf96be0c70c7624801f9234d23d53294/03af4f0b8f512e29de1d91bd47129909129fe550/issues?groups=code_patterns%3A4Phzg6k7%3Af1,code_patterns%3A5ZdRTEsh%3Af6,code_patterns%3Akv4o9Hx2,code_patterns%3A7I7lGdCx%3Af3&severity=4>
>> Functionality
>> 
>> Works as expected
>> Fix the loading of the first bucket-list
>> Fix token expiry
>> Testing
>> 
>> Good: 100% test coverage.
>> UI/UX
>> 
>> Need a favicon
>> Fix session expiry to be more usable
>> Call-to-action button
>> "Item Status" => "Done"
>> The rest of the nitpicks here <https://drive.google.com/a/andela.co/folderview?id=0B8IqKJACNaFOUzk3Y2hRNFlEbDA&usp=sharing>
>> —
>> Reply to this email directly or view it on GitHub <https://github.com/andela-lkabui/checkpoint3/pull/5#issuecomment-185597691>.
>> 
> 
> Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
> Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>
> 
> "The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
> See what The Wall Street Journal <http://snip.ly/u9AI> has to say.
> 

Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>

"The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
See what The Wall Street Journal <http://snip.ly/u9AI> has to say.

andela-lkabui commented:
Hi Nandaa,

I hope you are well today.

This is just a follow up on my request to submit checkpoint3 after the changes from the review.

Please advise on what time would work best for you.

Regards.
> On 25 Feb 2016, at 12:08, Lewis Maina <lewis.maina@andela.com> wrote:
> 
> I’ve just realised I did not send the link. Please find my repo here <https://github.com/andela-lkabui/checkpoint3/pull/12>.
>> On 25 Feb 2016, at 11:34, Lewis Maina <lewis.maina@andela.com <mailto:lewis.maina@andela.com>> wrote:
>> 
>> Good Morning Nandaa,
>> 
>> I trust your morning is great.
>> 
>> Having had the checkpoint3 review last week, I have addressed the nitpicks we identified during our session and I am ready to present the improvements.
>> 
>> I am hoping we can have some time before the end of this week (today if possible) to review the changes.
>> 
>> Please tell me if this works for you.
>>> On 18 Feb 2016, at 11:34, Anthony Nandaa <notifications@github.com <mailto:notifications@github.com>> wrote:
>>> 
>>> README / Comments
>>> 
>>> The README is detailed and okay.
>>> You could have a proper name for your app, other than Checkpoint3
>>> Directory Structure
>>> 
>>> Separate the API (DRF/Django app) and the Django app (front-end).
>>> Git / Commit Messages
>>> 
>>> Some of your git commit messages was not well wrapped, see how to fix this here <http://stackoverflow.com/questions/11023194/git-vim-automatically-wrap-long-commit-messages>
>>> Code Quality
>>> 
>>> Generally okay except for a few severe issues raised by Quantified Code <https://www.quantifiedcode.com/app/project/gh:andela-lkabui:checkpoint3/diff/1591b1f8cf96be0c70c7624801f9234d23d53294/03af4f0b8f512e29de1d91bd47129909129fe550/issues?groups=code_patterns%3A4Phzg6k7%3Af1,code_patterns%3A5ZdRTEsh%3Af6,code_patterns%3Akv4o9Hx2,code_patterns%3A7I7lGdCx%3Af3&severity=4>
>>> Functionality
>>> 
>>> Works as expected
>>> Fix the loading of the first bucket-list
>>> Fix token expiry
>>> Testing
>>> 
>>> Good: 100% test coverage.
>>> UI/UX
>>> 
>>> Need a favicon
>>> Fix session expiry to be more usable
>>> Call-to-action button
>>> "Item Status" => "Done"
>>> The rest of the nitpicks here <https://drive.google.com/a/andela.co/folderview?id=0B8IqKJACNaFOUzk3Y2hRNFlEbDA&usp=sharing>
>>> —
>>> Reply to this email directly or view it on GitHub <https://github.com/andela-lkabui/checkpoint3/pull/5#issuecomment-185597691>.
>>> 
>> 
>> Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
>> Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>
>> 
>> "The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
>> See what The Wall Street Journal <http://snip.ly/u9AI> has to say.
>> 
> 
> Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
> Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>
> 
> "The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
> See what The Wall Street Journal <http://snip.ly/u9AI> has to say.
> 

Lewis Kabui / Andela.com <http://andela.com/?utm_source=andela&utm_medium=email&utm_campaign=signature> 
Software Developer / +254726348287 / @Andela <http://bit.ly/andelatwitter>

"The startup that's harder to get into than Harvard" - CNN <http://snip.ly/HfId>
See what The Wall Street Journal <http://snip.ly/u9AI> has to say.
